### ISSUE TITLE
Memory error when computing EER or minDCF score

### ISSUE THREAD

**Issue by zengchang233** [speechbrain/speechbrain#1387]
## Error information

```
speechbrain.core - Exception:
Traceback (most recent call last):
  File "speaker_verification_plda.py", line 384, in <module>
    eer, min_dcf = verification_performance(scores_plda)
  File "speaker_verification_plda.py", line 146, in verification_performance
    eer, th = EER(torch.tensor(positive_scores), torch.tensor(negative_scores))
  File "/home/smg/apps/anaconda3/envs/pytorch-light/lib/python3.8/site-packages/speechbrain/utils/metric_stats.py", line 448, in EER
    positive_scores = torch.cat(
RuntimeError: [enforce fail at CPUAllocator.cpp:68] . DefaultCPUAllocator: can't allocate memory: you tried to allocate 479513191100 bytes. Error code 12 (Cannot allocate memory)
```
## Data

I used CNCeleb dataset to train a speaker verification model and evaluate the model according to the official CNCeleb testing protocal which contains 3.7M trials (196 enrollment speakers and 17777 testing utterances, the total number of trials is 196 * 17777). The trials number is much huger than VoxCeleb. 

No matter calculating plda score or cosine score, the code throws the above exception. It seems that the program eats too much memory (my machine has 252GB memory). But if I store the score to a file and compute EER by using kaldi or the below code, it is OK.

```
def compute_eer(y_true, y_pred):
    fpr, tpr, threshold = roc_curve(y_true, y_pred, pos_label = 1)
    eer = brentq(lambda x: 1. - x - interp1d(fpr, tpr)(x), 0., 1.)
    threshold = interp1d(fpr, threshold)(eer)
    return eer, threshold
```

**Comment by zengchang233** [comment]
Some package I used to compute EER of above code

```
from scipy.optimize import brentq
from scipy.interpolate import interp1d
from sklearn.metrics import roc_curve
```

**Comment by anautsch** [comment]
Hi @zengchang233 thanks for notifying! 

For the amount of scores, the processing method might indeed need a revision. For the requested `479513191100 bytes` (400+ GB), 252 GB memory will not suffice. We will need some score file architecture/design to handle lots of scores (like 3.7M). We are under-resourced right now (can't draft a timeline before September).

**Comment by zengchang233** [comment]
@anautsch Thanks for quick relpy! I am looking forward to the new code!

**Comment by anautsch** [comment]
Well, it's community-driven - 

If you or someone else has a good idea how to improve this issue, please open up a PR and we can move from there. The "projects" (rather discussion topics) assignment is more to not loose track of this issue in the meantime - in case fixing of other, related, PRs and issues is attempted. At the moment, a timeline for new code cannot be estimated.

**Comment by zengchang233** [comment]
I see. Actually as I mentioned, I can use other third-party packages to compute EER much faster than current code. But I am not sure whether it conflicts with the design of speechbrain which implements almost functions by only PyTorch.

**Comment by AlmogAmiga** [comment]
> Well, it's community-driven -
> 
> If you or someone else has a good idea how to improve this issue, please open up a PR and we can move from there. The "projects" (rather discussion topics) assignment is more to not loose track of this issue in the meantime - in case fixing of other, related, PRs and issues is attempted. At the moment, a timeline for new code cannot be estimated.

Hi, I've been facing a similar issue myself when I was trying to get EER on my dataset (SRE16 from Kaldi).
This happens because the code saves each FAR and FRR value inside an array (torch.tensor) and thus creates many values that take up all of the memory. My solution to that issue which I've applied only to the EER function and I'm sure that can be 
also applied to the minDCF function is to save the FAR[min_index] and FRR[min_index] values in separate variables and update them each time.

### This is my version for the EER function:
`def EER(positive_scores, negative_scores):
    """Computes the EER (and its threshold).

    Arguments
    ---------
    positive_scores : torch.tensor
        The scores from entries of the same class.
    negative_scores : torch.tensor
        The scores from entries of different classes.

    Example
    -------
    >>> positive_scores = torch.tensor([0.6, 0.7, 0.8, 0.5])
    >>> negative_scores = torch.tensor([0.4, 0.3, 0.2, 0.1])
    >>> val_eer, threshold = EER(positive_scores, negative_scores)
    >>> val_eer
    0.0
    """

    # Computing candidate thresholds
    thresholds, _ = torch.sort(torch.cat([positive_scores, negative_scores]))
    thresholds = torch.unique(thresholds)

    # Adding intermediate thresholds
    interm_thresholds = (thresholds[0:-1] + thresholds[1:]) / 2
    thresholds, _ = torch.sort(torch.cat([thresholds, interm_thresholds]))

    min_index = 1
    final_FRR = 0
    final_FAR = 0

    for i, cur_thresh in enumerate(thresholds):
        pos_scores_threshold = positive_scores <= cur_thresh
        FRR = (pos_scores_threshold.sum(0)).float() / positive_scores.shape[0]
        del pos_scores_threshold

        neg_scores_threshold = negative_scores > cur_thresh
        FAR = (neg_scores_threshold.sum(0)).float() / negative_scores.shape[0]
        del neg_scores_threshold

        # Finding the threshold for EER
        if (FAR - FRR).abs().item() < abs(final_FAR - final_FRR) or i == 0:
           min_index = i
           final_FRR = FRR.item()
           final_FAR = FAR.item()

    # It is possible that eer != fpr != fnr. We return (FAR  + FRR) / 2 as EER.
    # EER = (FAR[min_index] + FRR[min_index]) / 2
    EER = (final_FAR + final_FRR) / 2


    return float(EER), float(thresholds[min_index])`

**Comment by Adel-Moumen** [comment]
Fixed in #1468.
